### DISCUSSION TITLE
Bug: KeyError rather than TypeError in LocStream for python >=3.12

### DISCUSSION THREAD

**Discussion by xylar** [esmf-org/esmf#313] [ESMPy] [answered]
### Requirements

- [X] Reviewed [ESMPy Documentation](https://earthsystemmodeling.org/esmpy/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### Affiliation(s)

LANL

### ESMF Version

v8.7.0 (and earlier)

### Issue

I'm trying to build ESMPy 8.7.0 on conda-forge:
https://github.com/conda-forge/esmpy-feedstock/pull/85
In the tests, I am seeing a KeyError:
```
___________________________ TestLocStream.test_slice ___________________________

self = <esmpy.test.test_api.test_locstream.TestLocStream testMethod=test_slice>

    @pytest.mark.skipif(pet_count()!=1, reason="test must be run in serial")
    def test_slice(self):
        locstream = LocStream(5, name="Test LocStream")
    
>       locstream2 = locstream[0:2]

src/esmpy/test/test_api/test_locstream.py:95: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LocStream:
    name = 'Test LocStream' 
    lower_bounds = array([0], dtype=int32) 
    upper_bounds = array([5], dtype=int32) 
    keys = dict_items([]) 

slc = slice(0, 2, None)

    def __getitem__(self, slc):
        # initialize slc_ls
        slc_ls = slc
    
        # check that this is actually a slicing operation and not just regular item retrieval
        if not isinstance(slc, str):
            # parallel slicing is not yet enabled (collective operation)
            if pet_count() > 1:
                raise SerialMethod
            # re-initialize slc_ls
            slc_ls = get_formatted_slice(slc, self.rank)
    
        # slice at will
        try:
>           ret = super(LocStream, self).__getitem__(slc_ls)
E           KeyError: slice(0, 2, None)

src/esmpy/api/locstream.py:130: KeyError
``` 
The cause of this is that `slice()` objects are hashable in Python >=3.12, whereas they were not in Python <3.12, and raised a `TypeError`.  I believe the fix would be to change the `try`/`catch` to:
``` python
       # slice at will
       if not isinstance(slc_ls, slice):
           ret = super(LocStream, self).__getitem__(slc_ls)
       else:
           # as of Python 3.12, slice objects are hashable so don't
           # raise a TypeError
           ret = self.copy()
...
```
This at least seems to have allowed all tests to pass.

### Autotag

@billsacks

**Comment by xylar** [comment]
In general, it seems like you all should probably either be testing with python 3.12 and 3.13 or should explicitly constrain python to be below these versions.

**Comment by billsacks** [comment]
Thank you for reporting this @xylar . I'll take a look soon, and will also review the python versions that we're testing with.

**Comment by billsacks** [comment, marked as answer]
@xylar - sorry it took me a while to get to this, but I have now brought in your suggested change: 338cd064a2db2966f22d91188077a9e70906a6d2. I looked this over pretty carefully and agree with the change you proposed. As noted in that commit, I think there are a number of possible cases where the new code will behave differently from the old code, but as far as I can tell, all of those cases would have raised exceptions in the old code, and now will just raise different exceptions, so I'm not too concerned.

Part of what held me up was that I ran into different errors while testing with python 3.12 on my Mac: I was getting Segmentation Faults in two places. I also fixed this issue today: 6d5fdfe3bab588b11da023d8a0efc9cb62a2c777.

I will now add nightly testing with python3.12 and python3.13.

Thank you again for reporting this and for providing the fix!

**Reply by xylar** [reply]
Thanks @billsacks!

**Comment by billsacks** [comment]
@xylar - also FYI just in case you run into this: In updating our ESMPy test environments to test more recent versions of python as well as other packages, I ran into an issue with recent versions of numpy: https://github.com/esmf-org/esmf/issues/329. I'll work on a fix for that as well, so that we can hopefully have support for the latest python & numpy versions in the upcoming 8.8 release.

**Reply by xylar** [reply]
That sounds good.  And it sounds like maybe it would be good to set constraints on esmpy dependencies to explicitly exclude numpy 3 and python 3.14 until they are tested, based on this experience.

**Reply by billsacks** [reply]
Good idea. You'll have to excuse my ignorance, though: Would the best way to do this be via https://github.com/esmf-org/esmf/blob/develop/src/addon/esmpy/pyproject.toml?

**Reply by billsacks** [reply]
(I'm particularly wondering if that will be sufficient for constraints on the conda side, or if something else would need to be done in addition to that.)

**Reply by xylar** [reply]
Yes, exactly. We use that file as our instructions from you the developers when we updated the conda package.

Ideally, there would be an upper bound on python on line 10 and both lower and upper bounds numpy on line 13.

**Reply by billsacks** [reply]
@xylar - I was thinking of adopting https://scientific-python.org/specs/spec-0000/, which recommends supporting the last 3 years of python versions and the last 2 years of external packages like numpy. This would imply currently supporting python 3.11, 3.12 and 3.13; and numpy 1.25, 1.26, 2.0 and 2.1. In practice, to avoid a combinatoric explosion of testing, I'll probably set up nightly testing with python 3.11 + numpy 1.25, python 3.12 + numpy 2.0, and python 3.13 + numpy 2.1. Any thoughts on whether this feels reasonable based on your experience maintaining the ESMPy conda package or others? I'm definitely open to suggestions here.

**Reply by billsacks** [reply]
Particularly note that this proposal involves bumping the supported minimum python version from 3.7 to 3.11.

Hmmm, as I think about this, I'm concerned that this might be too strict of a policy given our support for operational centers that can be rather slow to adopt new versions. So maybe for now I'll extend this to 5 years of python versions (back to 3.9) and 3 years of external packages (back to numpy 1.22). I may not set up testing with all of these versions, but would at least test with the earliest, latest and some midpoint of those supported versions.

Again, happy to hear any thoughts you might have.

**Reply by billsacks** [reply]
Actually, some further-evolving thinking: I see that xESMF supports [python >= 3.8](https://github.com/pangeo-data/xESMF/blob/feebcfc11147eb00e3256924f0fded1485b7df60/pyproject.toml#L16) and [numpy >= 1.16](https://github.com/pangeo-data/xESMF/blob/feebcfc11147eb00e3256924f0fded1485b7df60/requirements.txt#L5). I'll see if we can support the same for now, though with a hope of restricting this somewhat more moving forward.
- Update: the earliest numpy I can get from conda is 1.19, so that's the earliest I'll test with / support.

I also see recommendations ([here](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#requires-python-upper-bounds) and [here](https://packaging.python.org/en/latest/guides/dropping-older-python-versions/#specify-the-version-ranges-for-supported-python-distributions)) to *not* specify upper limits on versions, so I'm thinking I'll just stick with minimum versions for now unless you have strong feelings otherwise.

**Reply by xylar** [reply]
@billsacks, I agree, jumping to python 3.11 as the lower bound is too aggressive.

Conda-forge supports python 3.9 to 3.13 at the moment.

I would suggest testing as many python versions as you can afford (but definitely the earliest and latest) as well as the earliest and latest numpy versions (i.e. at least one numpy 1 and one numpy 2 version).

I don't see too much value in supporting python 3.8 at this point as I understand it to have reached end of life.  But best to coordinate this with xESMF.

**Reply by billsacks** [reply]
Thanks a lot for your thoughts, @xylar , and for bearing with me in my ramblings last night!

For now I have settled on testing:
- python3.8 + numpy1.19
- python3.11 + numpy1.26
- python3.13 + numpy2.1

I'll talk with xESMF folks about bumping their minimum supported versions to at least python3.9 and maybe a slightly later version of numpy. I'll also consider adding more tested version combinations relatively soon.

**Reply by xylar** [reply]
Sounds good!
